File: LovePlay-UIKit/LovePlay/Class/Zone/View/DiscussDetailWebCell.m
//
//  DiscussDetailWebCell.m
//  LovePlay
//
//  Created by weiying on 2017/2/7.
//  Copyright © 2017年 Yuns. All rights reserved.
//

#import "DiscussDetailWebCell.h"

@interface DiscussDetailWebCell ()<UIWebViewDelegate>
//UI
@property (nonatomic, strong) UIWebView *webView;
//Data
@property (nonatomic, strong) NSString *htmlBody;
@property (nonatomic, assign) NSInteger webViewHeight;
@property (nonatomic, copy) webViewFinishLoadBlock finishLoadBlock;
@end

@implementation DiscussDetailWebCell

+ (instancetype)cellWithTableView:(UITableView *)tableView
{
    static NSString *ID = @"DiscussDetailWebCell";
    DiscussDetailWebCell *cell = [tableView dequeueReusableCellWithIdentifier:ID];
    if (cell == nil) {
        cell = [[DiscussDetailWebCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:ID];
    }
    return cell;
}

- (void)webViewDidFinishLoadBlock:(webViewFinishLoadBlock)finishLoadBlock
{
    _finishLoadBlock = finishLoadBlock;
}

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        self.selectionStyle = UITableViewCellSelectionStyleNone;
        [self addSubviews];
    }
    return self;
}

#pragma mark - private
- (void)addSubviews
{
    [self.contentView addSubview:self.webView];
}

#pragma mark - public
- (void)setupHtmlBoby:(NSString *)htmlBody
{
    _htmlBody = htmlBody;
    if (htmlBody.length > 0) {
        NSURL *cssURL = [NSURL fileURLWithPath:[[NSBundle mainBundle]pathForResource:@"News" ofType:@"css"]];
        [_webView loadHTMLString:[self handleWithHtmlBody:htmlBody] baseURL:cssURL];
    }
}

- (NSString *)handleWithHtmlBody:(NSString *)htmlBody
{
    NSString *html = [htmlBody stringByReplacingOccurrencesOfString:@"\t" withString:@""];
    NSString *cssName = @"News.css";
    NSMutableString *htmlString =[[NSMutableString alloc]initWithString:@"<html>"];
    [htmlString appendString:@"<head><meta charset=\"UTF-8\">"];
    [htmlString appendString:@"<link rel =\"stylesheet\" href = \""];
    [htmlString appendString:cssName];
    [htmlString appendString:@"\" type=\"text/css\" />"];
    [htmlString appendString:@"</head>"];
    [htmlString appendString:@"<body>"];
    [htmlString appendString:html];
    [htmlString appendString:@"</body>"];
    [htmlString appendString:@"</html>"];
    return htmlString;
}

#pragma mark - webView delegate
- (void)webViewDidFinishLoad:(UIWebView *)webView
{
    CGFloat webViewHeight = [[webView stringByEvaluatingJavaScriptFromString:@"document.body.offsetHeight"] floatValue] + 10;
    _webViewHeight = webViewHeight;
    
    if (_finishLoadBlock) {
        _finishLoadBlock(webViewHeight);
    }
}

#pragma mark - layout
- (void)layoutSubviews
{
    [super layoutSubviews];
    
    if (_webViewHeight > 0) {
        _webView.height = _webViewHeight;
    }
}

#pragma mark - other
- (void)dealloc
{
    _webView.delegate = nil;
    _webView = nil;
}

#pragma mark - setter / getter
- (UIWebView *)webView
{
    if (!_webView) {
        UIWebView *webView = [[UIWebView alloc] initWithFrame:CGRectMake(0, 0, kScreenWidth, 1)];
        webView.backgroundColor = [UIColor whiteColor];
        webView.scalesPageToFit = NO;
        webView.delegate = self;
        webView.scrollView.bounces = NO;
        [webView setAutoresizingMask:UIViewAutoresizingNone];
        [webView.scrollView setScrollEnabled:NO];
        [webView.scrollView setScrollsToTop:NO];
        _webView = webView;
    }
    return _webView;
}
@end

File: LovePlay-Swift3/LovePlay/Class/Zone/View/DiscussDetailWebCell.swift
//
//  DiscussDetailWebCell.swift
//  LovePlay
//
//  Created by Yuns on 2017/6/11.
//  Copyright © 2017年 yuns. All rights reserved.
//

import UIKit

class DiscussDetailWebCell: UITableViewCell, UIWebViewDelegate {

    var finishBlock : WebViewDidFinishLoadBlock?
    
    class func cellWithTableView(tableView : UITableView) -> DiscussDetailWebCell {
        let ID = NSStringFromClass(self)
        var cell : DiscussDetailWebCell? = tableView.dequeueReusableCell(withIdentifier: ID) as? DiscussDetailWebCell
        if cell == nil {
            cell = DiscussDetailWebCell(style: .default, reuseIdentifier: ID)
        }
        return cell!
    }
    
    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        self.selectionStyle = .none
        self.addSubViews()
        self.snp_subViews()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    
    // MARK: - private
    private func addSubViews() {
        self.contentView.addSubview(self.discussWebView)
    }
    
    private func snp_subViews() {
        self.discussWebView.snp.makeConstraints { (make) in
            make.edges.equalToSuperview()
        }
    }
    
    fileprivate var _htmlBody : String?
    var htmlBody : String? {
        set {
            _htmlBody = newValue
            
            var html : String = (_htmlBody?.replacingOccurrences(of: "\t", with: ""))!
            let cssName = "News.css"
            html = "<html>" +
                "<head><meta charset=\"UTF-8\">" +
                "<link rel =\"stylesheet\" href =" +
                cssName +
                "type=\"text/css\" />" +
                "</head>" +
                "<body>" +
                html +
                "<body>" +
            "<html>"
            let url : URL = URL.init(fileURLWithPath: Bundle.main.path(forResource: cssName, ofType: nil)!)
            self.discussWebView.loadHTMLString(html, baseURL: url)
        }
        
        get {
            
            return _htmlBody
        }
    }
    
    // MARK: - public
    func webViewDidFinishLoadBlock(finishLoadBlock : @escaping WebViewDidFinishLoadBlock) {
        self.finishBlock = finishLoadBlock
    }
    
    
    // MARK: - delegate
    func webViewDidFinishLoad(_ webView: UIWebView) {
        let jsCode = "document.body.offsetHeight"
        
        let webViewHeight = Float((webView.stringByEvaluatingJavaScript(from: jsCode))!)

        let webHeight = CGFloat(webViewHeight!) + 10.0
        // update UI
        self.discussWebView.height = webHeight
        self.discussWebView.snp.updateConstraints { (make) in
            make.edges.equalToSuperview()
        }
        
        if finishBlock != nil {
            finishBlock!(webHeight)
        }
    }

    // MARK: - setter / getter
    lazy var discussWebView : UIWebView = {
        let webView = UIWebView(frame: CGRect(x: 0, y: 0, width: self.contentView.width, height: 1))
        webView.delegate = self
        webView.scrollView.isScrollEnabled = false
        return webView
    }()
}
